### PR TITLE
[modules] Dependent command packages can now be built with ko 

### PR DESCRIPTION
**Pull request by dprotaso**
Thus if you have a (in)direct command package as a dependency
say `myhost.com/go/package/cmd/run` you can now publish this
with the following ko command

```bash
$  ko publish myhost.com/go/package/cmd/run
```

This follows go's module/vendor semantics. Thus go1.14 by default will use vendored modules. With go1.13 you need the explicit flags to use vendored deps

```bash
$  GOFLAGS="-mod=vendor" ko publish myhost.com/go/package/cmd/run
```

Subsequently in yaml files you can drop the superfluous prefixes

ie. this (in a non-module world)  

```yaml
apiVersion: batch/v1
kind: Job
spec:
  containers:
  - name: vegeta
    image: knative.dev/serving/vendor/github.com/tsenart/vegeta
```
becomes this in the module world

```yaml
apiVersion: batch/v1
kind: Job
spec:
  containers:
  - name: vegeta
    image: github.com/tsenart/vegeta
```

### CODE REVIEW

**Comment by dprotaso** [comment]
cc https://github.com/google/ko/issues/152

File: pkg/build/gobuild.go [hunk 1]
```diff
@@ -79,34 +79,73 @@ func (gbo *gobuildOpener) Open() (Interface, error) {
...
 //
 // Related: https://github.com/golang/go/issues/26504
-func moduleInfo() *modInfo {
-	output, err := exec.Command("go", "list", "-mod=readonly", "-m", "-json").Output()
+func moduleInfo() (*modules, error) {
+	modules := modules{
+		deps: make(map[string]*modInfo),
+	}
+
+	// TGDO - optimize
```
Comment (ImJasonH) [inline pkg/build/gobuild.go:102]: What's this?

**Comment by ImJasonH** [comment]
I'm not sure I understand what this is trying to achieve. Is the effect that you can `ko build` a non-`package main` package?

File: pkg/build/gobuild.go (hunk 1 above)
Comment (dprotaso) [reply pkg/build/gobuild.go]: see updated comment

**Comment by dprotaso** [comment]
whoops needs a rebase - one minute

**Comment by dprotaso** [comment]
Failure in CI is go1.12 

> pkg/build/gobuild.go:134: Errorf format %w has unknown verb w

1.12 is EOL so I made https://github.com/google/ko/pull/156

**Comment by mattmoor** [comment]
@ImJasonH If I'm reading the description correctly, this is the `ko publish` complement of my change.

File: pkg/build/gobuild.go [hunk 2]
```diff
@@ -75,38 +81,80 @@ func (gbo *gobuildOpener) Open() (Interface, error) {
...
+		err := dec.Decode(&info)
+		if err == io.EOF {
+			// all done
+			break
+		}
+
+		modules.deps[info.Path] = &info
+
+		if info.Main {
+			modules.main = &info
```
Comment (mattmoor) [inline pkg/build/gobuild.go:130]: couldn't we hit N of these?  Which one wins?

**Comment by mattmoor** [comment]
Ok, so reading through the change, I want to make sure I understand...

With my change, a vendored main module `foo` will work if it is written as `ko://foo`, but what you are doing is dumping all of the referenced modules, so that just `foo` still works.

Since in `ko publish` qualify local import turns `./vendor/foo` into `foo` with modules, this unblocks letting `ko publish ./vendor/foo` or `ko publish foo` from working.

@dprotaso is that right?

File: pkg/build/gobuild.go (hunk 2 above)
Comment (dprotaso) [reply pkg/build/gobuild.go:130]: I believe there's only one

**Comment by dprotaso** [comment]
> this unblocks letting ko publish ./vendor/foo or ko publish foo from working.

This unblocks `ko publish foo`
This _does not unblock_ `ko publish ./vendor/foo`

File: pkg/build/gobuild.go (hunk 2 above)
Comment (dprotaso) [reply pkg/build/gobuild.go:130]: main != `package main` but specifically "main module" = root module

**Comment by mattmoor** [comment]
I wonder, would simply having `qualifyLocalImport` slap a `ko://` on the front of what it produces work?

File: pkg/build/gobuild.go (hunk 2 above)
Comment (mattmoor) [reply pkg/build/gobuild.go:130]: that's not confusing *at all* 🙄

**Comment by dprotaso** [comment]
🤷‍♂️

**Comment by mattmoor** [comment]
This just makes me want to deprecated non-`ko://` faster, but it seems like an improvement of the status quo.  thanks for chasing this down.

**Pull request merged**